server/publications/teams.js

```diff
@@ -12,9 +12,9 @@ Meteor.publish('admin.teams', function() {
     inPerson: 1,
     hasBegun: 1,
     owner: 1,
-    'puzzles.start': 1,
-    'puzzles.end': 1,
-    'puzzles.name': 1,
+    prize_ineligible: 1,
+    EMERGENCY_LOCK_OUT: 1,
+    puzzles: 1,
```

Comment:
I'm a little uncomfortable with this change. This exposes some more "sensitive" information like download URL, # hints taken, and puzzle answer. I think in theory someone could sniff the data when the load a page like the leaderboard and then get lots of secret info. If there's a new field you need, can you add it explicitly? Or is there some other reason for this kind of change? Otherwise, I'd like to go back to specifying specific puzzle fields

Reply:
There is a specific reason, and the alternative approaches are too expensive for us to implement.

This is another instance of Meteor's poor implementation of client-side collection storage. When one includes only particular fields of puzzles in a projection that is tracking the team collection, potentially *only* those fields will be available in the client--even if there is another publication that asks for all the fields. Internally, Meteor caches a record in the client the first time some publication sends it. After that, Meteor will not update that record until it changes in the database. Meteor explicitly does not attempt to merge in additional fields once the record has been cached; they assert that this is a design feature. The implication is that the contents of the cached collection in the client is indeterminate, depending on which tracker or publication first retrieved the record. In other words, if a client *can* access a limited view of a record, sound design *must* assume that only the fields in that limited view are available.

This causes many subtle bugs. In particular, it was causing an exception in the getHintsTaken() method in lib/imports/puzzle-helpers.js, which is called from all over the code. It would do this intermittently, depending on which actions had been done a few steps before. If the right previous actions had accessed the whole record, it worked. If not, it raised an exception.

I have found many similar bugs that were being caused by Meteor's caching design. For example, the admin team dialog did not show whether a team was locked out or not because sometimes that field was not getting sent to the client.

If we want to avoid sending some of the information to the client, as you suggest, then we have to bifurcate the admin.teams publication, creating one that is used only when a client is logged in as non-admin, and one that is used only when a client is logged in as admin. The non-admin publication would need to error out if a client logged in as admin tried to use it. When an admin is logged in, their client must never use a publication that will send records with necessary fields projected out.

That is doable, but not with the resources and time we have available. We would be at minimum duplicating the player-visible leaderboard and various teams pages to have parallel admin-visible versions. We would also need to trawl through all of the client code to ensure that an admin only calls admin-appropriate publications and similar for non-admins, and get enough testing done to ensure that all the induced bugs were caught. I would guess that is minimum two to three weeks at more than half time to get it done. I don't have the hours available for that, and I'm guessing that you don't either. (There are other alternatives, like handling the puzzle information differently in a team record, or switching to an infrastructure that doesn't have the design problems that Meteor does; those seem clearly far more expensive.)

I agree that this change sends more information to a non-admin client than we would ideally want. However, the puzzle hunt does not deal with any truly sensitive information; there is no information in it that rates anything other than FIPS confidentiality low. If someone were so motivated to sniff the data, the consequences would be minor at best. (Ideally I wouldn't make that argument, because I would rather have a clean design.)

Given the cost versus benefit tradeoff, it doesn't make sense to try to fix this.